### PR TITLE
do not try an exponential number of package names

### PR DESCRIPTION
**Pull request by Eh2406**
re #11934, and as discussed in the cargo team meeting, this changes the strategy to "the original, all underscore, and all dashes".

I was excessively proud of the `hyphen_combination_num` based implementation when I came up with it. But it's always been a hack. I'm glad to be the one to remove it.

### CODE REVIEW

File: src/cargo/sources/registry/mod.rs [hunk 1]
```diff
@@ -850,9 +850,17 @@ impl<'cfg> Source for RegistrySource<'cfg> {
...
-                for name_permutation in
-                    index::UncanonicalizedIter::new(&dep.package_name()).take(1024)
+                // For now we only try the canonical lysing `-` to `_` and vice versa.
+                // More advanced fuzzy searching become in the future.
+                for name_permutation in [
+                    dep.package_name().replace('-', "_"),
+                    dep.package_name().replace('_', "-"),
+                ]
+                .into_iter()
                 {
```
Comment (weihanglo) [inline src/cargo/sources/registry/mod.rs:860]: We don't need [`into_iter` for array](https://doc.rust-lang.org/edition-guide/rust-2021/IntoIterator-for-arrays.html) anymore :)



```suggestion
                ] {
```

**Review by weihanglo — APPROVED** [review]
Looks good!

Just give a data point: The number of crate names on crates.io using a mix of dash and underscore grows from 623(<https://github.com/rust-lang/cargo/issues/11934#issuecomment-1496821761>) to 639 (2023-05-05-020041).

**Comment by weihanglo** [comment]
Feel free to `r=weihanglo` or if you want to do an fcp.

**Comment by weihanglo** [comment]
Is it worth a <https://github.com/rust-lang/cargo/labels/relnotes> for putting it in Compatibility Notes in RELEASE.md?

**Comment by qm3ster** [comment]
I feel like the easiest to make mistake for mixed cases is inversion (`my-rice_your-peas` vs `my_rice-your_peas`), so perhaps that is one last case that needs to be included.
I initially thought it may actually be the only thing worth including, since if a user is using both symbols in one name they probably have different semantics in mind, while the substitution will also cover both non-mixed inversions.
However I now realize the crate author may still canonicalize both semantics to one or the other, so all are needed.

**Pull request merged**